Add retry policy to Remote Query Executor to be able to retry failed leaf queries

seandhaynes opened the pull request:
This is a continuation of https://github.com/ClickHouse/ClickHouse/pull/71059, including addressing feedback kindly provided in reviews from @devcrafter . I don't have permission to push to the original branch and it's significantly out of date, so I've started a new one here.

As ClickHouse clusters scale and you begin to have multiple
shards with many replicas, the probability of 1 of those nodes
saturated, busy or not in a condition to serve queries
increases.

This can often mean that a Distributed query say across dozens
of shards, might fail just because of a single node. This means
all the work and resources we've used on other shards is effectively
thrown away, and we return an error to the client which can choose
to retry the whole query if that want to.

This is quite wasteful, and can often lead to cascading values if
clusters are busy.

This commit adds a new setting:

`distributed_shard_retry_count` which defaults to 0 to effectively turn
retries off for backwards compatible behaviour.

If this setting is > 0 and a leaf node in a distributed query returns
an error, we will retry the query on a different leaf node.

Co-authored-by: Christian Cendres
Co-authored-by: Anton Kozlov

### Changelog category (leave one):
- New Feature

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Add the ability retry certain errors in distributed queries

### Documentation entry for user-facing changes

- [ x] Documentation is written (mandatory for new features)


- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

distributed_shard_retry_count - the number of times a distributed query should be retried on a leaf node on the event it hits a retryable error.

seandhaynes committed:
Add the ability retry certain errors in distributed queries

As ClickHouse clusters scale and you begin to have multiple
shards with many replicas, the probability of 1 of those nodes
being saturated, busy or not in a condition to serve queries
increases.

This can often mean that a Distributed query say across dozens
of shards, might fail just because of a single node. This means
all the work and resources we've used on other shards is effectively
thrown away, and return an error to the client, which then can choose
to retry the whole query if that want to.

This is quite wasteful, and can often lead to cascading values if
clusters are busy.

This commit adds a new setting:

`distributed_shard_retry_count` which defaults to 0 to effectively turn
retries off for backwards compatible behaviour.

If this setting is > 0 and a leaf node in a distributed query returns
an error, we will retry the query on a different leaf node.

Co-authored-by: Christian Cendres
Co-authored-by: Anton Kozlov

devcrafter commented on src/QueryPipeline/RemoteQueryExecutor.cpp:
```diff
@@ -589,38 +610,50 @@ RemoteQueryExecutor::ReadResult RemoteQueryExecutor::readAsync()
         if (was_cancelled)
             return ReadResult(Block());
 
-        /// Cancel previous query and disconnect before retry.
```
I'd assume that we have to cancel query before throwing exception bellow

devcrafter commented on tests/integration/test_distributed_shard_retries/test.py:
```diff
@@ -0,0 +1,121 @@
+    result_value = int(result.split("\t")[0])
+    assert result_value == 999, f"Expected max(i) == 999, got {result_value}"
+
+    node1.query("SYSTEM FLUSH LOGS")
```
need to flush only `query_log`

devcrafter commented on src/Client/MultiplexedConnections.cpp:
```diff
@@ -541,6 +541,28 @@ void MultiplexedConnections::setAsyncCallback(AsyncCallback async_callback)
+            auto * pool_base = state.pool_entry.getPool();
+            if (auto * conn_pool = dynamic_cast<ConnectionPool*>(pool_base))
+            {
+                return ConnectionPoolPtr(conn_pool, [](IConnectionPool*){});
```
This looks like a hack. Can we do better?

seandhaynes replied:
Thanks for looking at this. I've actually completely removed all of this code now after finding what is in my opinion a much cleaner way by using NestedPool instances directly within RemoteQueryExecutor. Please do let me know what you think.

devcrafter reviewed (commented):
In general, looks ok, but can we come up with a better way to mark “failed” replica so it'll not be chosen again during retry

seandhaynes commented on src/QueryPipeline/RemoteQueryExecutor.cpp:
```diff
@@ -655,15 +739,55 @@ RemoteQueryExecutor::ReadResult RemoteQueryExecutor::processPacket(Packet packet
             got_exception_from_replica = true;
+
+            /// If the error is retryable and config suggests we should do so, reset query state and retry
+            const auto max_retries = context->getSettingsRef()[Setting::distributed_shard_retry_count];
```
I'm not sure of the value of aligning retries with the number of replicas in a shard.

The reason being is although we make an effort to increase the error count on the pool of replica that returned an error, it's incredibly valuable even retrying on the same node on  the exceptions we retry. Concurrency errors for example, we could retry that same node and there could be a concurrency slot available.

Please do let me know if you still think there's value in doing this though.

seandhaynes committed:
Improve how we get replica pool and add integration test

seandhaynes committed:
Move settings changes entry into latest release

devcrafter commented on src/QueryPipeline/RemoteQueryExecutor.cpp:
```diff
@@ -630,6 +672,57 @@ RemoteQueryExecutor::ReadResult RemoteQueryExecutor::restartQueryWithoutDuplicat
+    /// that the same query will succeed on a different replica.
+    return error_code == ErrorCodes::TOO_MANY_SIMULTANEOUS_QUERIES
+        || error_code == ErrorCodes::CANNOT_OPEN_FILE
+        || error_code == ErrorCodes::NETWORK_ERROR;
```
Can NETWORK_ERROR code be an error in ServerException? Currenly, the change handle only errors which received from remote server with exception packet. NETWORK_ERROR can happened only before processPacket() is called. The only possible case I can think of is distributed over distributed (but not sure what would happened in such case).

seandhaynes committed:
Address review feedback: simplify IConnections and harden retry path

Following review on PR #90724, move replica-pool tracking out of the
IConnections interface and address several correctness and clarity
comments on the shard-level retry path.

seandhaynes committed:
Merge remote-tracking branch 'origin/master' into shaynes/shard-level-retries

# Conflicts:
#	src/Common/FailPoint.cpp

seandhaynes committed:
Merge branch 'master' into shaynes/shard-level-retries

devcrafter commented on src/QueryPipeline/RemoteQueryExecutor.cpp:
```diff
@@ -590,23 +618,40 @@ RemoteQueryExecutor::ReadResult RemoteQueryExecutor::readAsync()
 }
 
 
+void RemoteQueryExecutor::resetQueryState()
+{
+    /// Reset state to retry
+    resent_query = true;
+    recreate_read_context = true;
+    sent_query = false;
+    was_cancelled = false;
+    received_data_from_replica = false;
+}
```
https://github.com/ClickHouse/ClickHouse/pull/99398 has been merged. It's time to update the PR

seandhaynes replied:
amazing, thanks @devcrafter . I'll update this PR now.
